onnxruntime/core/session/onnxruntime_c_api.cc

```diff
@@ -93,39 +67,113 @@ class LoggingWrapper : public ISink {
   void* logger_param_;
 };
 
+struct OrtEnv {
+ public:
+  struct LoggingManagerConstructionInfo {
+    OrtLoggingFunction logging_function{};
+    void* logger_param{};
+    OrtLoggingLevel default_warning_level;
+    const char* logid{};
+  };
+
+  static OrtEnv* GetInstance(const LoggingManagerConstructionInfo& lm_info, Status& status) {
+    std::lock_guard<OrtMutex> lock(m_);
+    if (!p_instance_) {
+      std::unique_ptr<Environment> env;
+      status = Environment::Create(env);
+      if (!status.IsOK()) {
+        return nullptr;
+      }
+
+      std::unique_ptr<LoggingManager> lmgr;
+      std::string name = lm_info.logid;
+      if (lm_info.logging_function) {
+        std::unique_ptr<ISink> logger = onnxruntime::make_unique<LoggingWrapper>(lm_info.logging_function,
+                                                                                 lm_info.logger_param);
+        lmgr.reset(new LoggingManager(std::move(logger),
+                                      static_cast<Severity>(lm_info.default_warning_level),
+                                      false,
+                                      LoggingManager::InstanceType::Default,
+                                      &name));
+      } else {
+        lmgr.reset(new LoggingManager(std::unique_ptr<ISink>{new CLogSink{}},
+                                      static_cast<Severity>(lm_info.default_warning_level),
+                                      false,
+                                      LoggingManager::InstanceType::Default,
+                                      &name));
+      }
+
+      p_instance_ = new OrtEnv(std::move(env), std::move(lmgr));
+    }
+    ++ref_count_;
+    return p_instance_;
+  }
+
+  static void Release(OrtEnv* env_ptr) {
+    if (!env_ptr) {
+      return;
+    }
+    std::lock_guard<OrtMutex> lock(m_);
```

Comment:
Is it possible to remove this lock? Because it increases the risk of having deadlock.